opusfilter/opusfilter.py

```diff
@@ -71,6 +74,108 @@ def dict_set(key, value, dictionary):
+    """
+    Decorator for parallelizing 'filter_data', 'score_data' and 'preprocess' methods.
+    This decorator will split "inputs" and "outputs" (or "output") into shareds and process them in parallel.
+    If the output file name is $name, the output file name will be $name.0, $name.1, ...
+    Finally, all the intermediate result will be merged into a single file $name. All the intermediate files will be deleted.
+    """
+
+    def __init__(self, extra_parameters):
+        self.extra_parameters = extra_parameters
+        self.func = None
+
+    def __call__(self, func):
+        self.func = func
+
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            return self.parallelize(*args, **kwargs)
+        return wrapper
+
+    @staticmethod
+    def _split(infiles, outfiles, n_jobs):
+        """split files into parts, and write them to temporary files"""
+        chunk_size = int(math.ceil(count_lines(infiles[0]) / n_jobs))
+
+        in_chunked_files = []
+        out_chunked_file = []
+        infileobjs = [file_open(infile) for infile in infiles]
+        for i, lines in enumerate(zip(*infileobjs)):
+            if i % chunk_size == 0:
+                intmpfiles = [f"{fname}.{str(i // chunk_size)}" for fname in infiles]
+                in_chunked_files.append(intmpfiles)
+                outtmpfiles = [f"{fname}.{str(i // chunk_size)}" for fname in outfiles]
+                out_chunked_file.append(outtmpfiles)
+                intmpfiles_objs = [file_open(fname, mode="w") for fname in intmpfiles]
+
+            for line, tmpfile in zip(lines, intmpfiles_objs):
+                tmpfile.write(line)
+        for tmpfile in intmpfiles_objs:
+            tmpfile.close()
+        return in_chunked_files, out_chunked_file
+
+    @staticmethod
+    def _merge(in_chunked_files, outfiles, out_chunked_files, limit):
+        """merge temporary files into final files and delete temporary files"""
+        for outfile, parts in zip(outfiles, zip(*out_chunked_files)):
+            with file_open(outfile, 'w') as out:
+                for i, line in enumerate(fileinput.input(parts)):
+                    out.write(line)
+                    if limit and i >= limit - 1:
+                        break
+            for part in parts:
+                os.remove(part)
+        for parts in in_chunked_files:
+            for part in parts:
+                os.remove(part)
+
+    def parallelize(self, obj, parameters, overwrite=False):
+        """Wrapper for parallelizing a function"""
+        # check if parameters are valid
+        if not isinstance(obj, OpusFilter):
+            raise TypeError("Decorator parallel can only be used with functions in OpusFilter.")
+        if "inputs" not in parameters:
+            obj.extra_parameters.update({"inputs"})
+        if "outputs" not in parameters and "output" not in parameters:
+            obj.extra_parameters.update({"outputs"})
+        n_jobs = parameters.pop('n_jobs', 1)
```

Comment:
I'm wondering whether this `n_jobs` should rather be a command-line option to `opusfilter` than a configuration option. E.g. if you run the same configuration on different machines with different number of cores, you wouldn't want to change the actual configuration. But there could also be cases where it makes sense to use many cores for one filtering task, but just one for another. Maybe a `default_n_jobs` attribute to the `OpusFilter` class, with default 1 that could be overridden first by a command line option and then by a configuration-specific setting?

However, this is a feature that I can add also later, so no need to do the changes for the first version. If you have any thoughts on this I'm happy to hear.